src/MongoDB.Driver/Linq/Processors/OutOfCurrentScopePrefixCollector.cs

```diff
@@ -0,0 +1,53 @@
+﻿/* Copyright 2019-present MongoDB Inc.
+*
+* Licensed under the Apache License, Version 2.0 (the "License");
+* you may not use this file except in compliance with the License.
+* You may obtain a copy of the License at
+*
+* http://www.apache.org/licenses/LICENSE-2.0
+*
+* Unless required by applicable law or agreed to in writing, software
+* distributed under the License is distributed on an "AS IS" BASIS,
+* WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+* See the License for the specific language governing permissions and
+* limitations under the License.
+*/
+
+using System.Collections.Generic;
+using System.Linq;
+using System.Linq.Expressions;
+using MongoDB.Driver.Linq.Expressions;
+
+namespace MongoDB.Driver.Linq.Processors
+{
+    /// <summary>
+    /// The expression visitor to collect all expression fields which have been defined out of the current scope.
+    /// </summary>
+    internal sealed class OutOfCurrentScopePrefixCollector : ExtensionExpressionVisitor
+    {
+        private readonly List<IHasOutOfCurrentScopePrefix> _outOfCurrentScopePrefixCollection = new List<IHasOutOfCurrentScopePrefix>();
+
+        /// <summary>
+        /// Collects the list of node information.
+        /// </summary>
+        /// <param name="expression">The expression.</param>
+        /// <returns>The result collection.</returns>
+        public static IEnumerable<IHasOutOfCurrentScopePrefix> Collect(Expression expression)
+        {
+            var parameterCollector = new OutOfCurrentScopePrefixCollector();
+            parameterCollector.Visit(expression);
+            return parameterCollector._outOfCurrentScopePrefixCollection.Distinct();
```

Comment:
Instead of having to call Distinct here why not make _outOfCurrentScopePrefixCollection a Set?

Reply:
if you are talking about `Set` as a c# property accessor, then as I see, it will not work since this accessor will control the only assignment a `new` value to `_outOfCurrentScopePrefixCollection` and will not control items assignment in the list variable. 
I guess I will just add a helper method to add only unique values to the collection.

Reply:
https://docs.microsoft.com/en-us/dotnet/api/system.collections.generic.hashset-1?view=netframework-4.7.2